processing/src/main/java/io/druid/segment/IndexMergerV9.java

```diff
@@ -149,97 +149,113 @@ public Metadata apply(IndexableAdapter input)
-        ioPeon, mergedDimensions, indexSpec
-    );
-    final ArrayList<ByteBufferWriter<ImmutableRTree>> spatialIndexWriters = setupSpatialIndexWriters(
-        ioPeon, mergedDimensions, indexSpec, dimCapabilities
-    );
-    makeInvertedIndexes(
-        adapters, progress, mergedDimensions, indexSpec, v9TmpDir, rowNumConversions,
-        nullRowsList, dimValueWriters, bitmapIndexWriters, spatialIndexWriters, dimConversions
-    );
+      /************ Create Inverted Indexes *************/
+      progress.progress();
+      final ArrayList<GenericIndexedWriter<ImmutableBitmap>> bitmapIndexWriters = setupBitmapIndexWriters(
+          ioPeon, mergedDimensions, indexSpec
+      );
+      final ArrayList<ByteBufferWriter<ImmutableRTree>> spatialIndexWriters = setupSpatialIndexWriters(
+          ioPeon, mergedDimensions, indexSpec, dimCapabilities
+      );
+      makeInvertedIndexes(
+          adapters, progress, mergedDimensions, indexSpec, v9TmpDir, rowNumConversions,
+          nullRowsList, dimValueWriters, bitmapIndexWriters, spatialIndexWriters, dimConversions
+      );
 
-    /************ Finalize Build Columns *************/
-    progress.progress();
-    makeTimeColumn(v9Smoosher, progress, timeWriter);
-    makeMetricsColumns(v9Smoosher, progress, mergedMetrics, metricsValueTypes, metricTypeNames, metWriters);
-    makeDimensionColumns(
-        v9Smoosher, progress, indexSpec, mergedDimensions, dimensionSkipFlag, dimCapabilities,
-        dimValueWriters, dimWriters, bitmapIndexWriters, spatialIndexWriters
-    );
+      /************ Finalize Build Columns *************/
+      progress.progress();
+      makeTimeColumn(v9Smoosher, progress, timeWriter);
+      makeMetricsColumns(v9Smoosher, progress, mergedMetrics, metricsValueTypes, metricTypeNames, metWriters);
+      makeDimensionColumns(
+          v9Smoosher, progress, indexSpec, mergedDimensions, dimensionSkipFlag, dimCapabilities,
+          dimValueWriters, dimWriters, bitmapIndexWriters, spatialIndexWriters
+      );
 
-    /************* Make index.drd & metadata.drd files **************/
-    progress.progress();
-    makeIndexBinary(
-        v9Smoosher, adapters, outDir, mergedDimensions, dimensionSkipFlag, mergedMetrics, progress, indexSpec
-    );
-    makeMetadataBinary(v9Smoosher, progress, segmentMetadata);
+      /************* Make index.drd & metadata.drd files **************/
+      progress.progress();
+      makeIndexBinary(
+          v9Smoosher, adapters, outDir, mergedDimensions, dimensionSkipFlag, mergedMetrics, progress, indexSpec
+      );
+      makeMetadataBinary(v9Smoosher, progress, segmentMetadata);
 
-    v9Smoosher.close();
-    ioPeon.cleanup();
-    FileUtils.deleteDirectory(v9TmpDir);
-    progress.stop();
+      v9Smoosher.close();
+      progress.stop();
 
-    return outDir;
+      return outDir;
+    } finally {
+      IOException e = null;
```

Comment:
Can this just be done via a `Closer`?

The `Closer` code patterns have taken into account making sure everything closes and errors get thrown correctly.